Enable RLE on OSSA

meg-gupta opened the pull request.

meg-gupta commented:
@swift-ci test

meg-gupta commented:
@swift-ci benchmark

eeckstein commented on lib/SILOptimizer/Transforms/RedundantLoadElimination.cpp:
```diff
@@ -700,6 +727,59 @@ SILValue BlockState::reduceValuesAtEndOfBlock(RLEContext &Ctx, LSLocation &L) {
+    return Copy;
+  };
+
+  if (isa<CopyValueInst>(TheForwardingValue)) {
```
This is a "pro-actively" created copy_value from processWrite or processRead, right?
Can you mention this in a comment here?

Wouldn't it be easier to not "pro-actively" create the copy_values, but just create all (including the first one) here?
Also, is it always sufficient to create the copy_value at the location of the forwarding value? What if the load to forward is outside a loop and the load to replace inside a loop? In this case a copy_value must be inserted in the loop.

meg-gupta replied:
Yes, it is the proactively created copy_value from processWrite and processRead. This can also include copy_value of struct_extract if the redundant load was a field load and the available dominating load was a load of its struct.

We can't insert the first copy_value here because we need it to be available before RLEContext::collectLocationValues. This is because it ends up calling SSAUpdater::getValueInMiddleOfBlock which can insert a phi arg of the loaded value which ends its lifetime in OSSA. And if we have any post dominating use of the loaded value, it will end up being a value after a consuming use. `rle_needscopy1` in `redundant_load_elim_ossa_complex.sil` shows this case

meg-gupta replied:
Thanks for pointing out if its sufficient to create a copy_value at the location of the forwarding value. I realized RLE on OSSA is not removing such a load. Added test` rle_redundantload_does_not_postdominate2` and `rle_redundantload_does_not_postdominate3` in `redundant_load_elim_ossa_complex.sil`. I'll investigate this further.

eeckstein replied:
I see, thanks

meg-gupta replied:
I updated the PR with the fix and tests. I am now using SILLoopAnalysis to determine if we need a copy at the redundant load.

eeckstein replied:
SILLoopAnalysis is not the right tool for this. There might be cycles in the CFG which are not natural loops.

meg-gupta replied:
Thanks for suggesting ValueLifetimeAnalysis as an alternative offline.

eeckstein commented on lib/SILOptimizer/Transforms/RedundantLoadElimination.cpp:
```diff
@@ -1523,6 +1662,64 @@ void RLEContext::processBasicBlocksForRLE(bool Optimistic) {
+    // If the LastUser does not end the lifetime of Val, we have to insert a
+    // destroy_value of Val
+    bool HasLifetimeEnd = false;
+    if (LastUser) {
```
simpler:
```
  if (End != &End->getParent()->front())
```

eeckstein commented on lib/SILOptimizer/Transforms/RedundantLoadElimination.cpp:
```diff
@@ -1523,6 +1662,64 @@ void RLEContext::processBasicBlocksForRLE(bool Optimistic) {
+/// when not already found
+static void fixupLifetimeOfValue(ValueLifetimeAnalysis &Vla, SILValue Val) {
+  ValueLifetimeAnalysis::Frontier ValFrontier;
+  Vla.computeFrontier(ValFrontier, ValueLifetimeAnalysis::DontModifyCFG);
```
computeFrontier with ValueLifetimeAnalysis::DontModifyCFG may fail in case of critical edges. You might want to use AllowToModifyCFG

meg-gupta replied:
I thought since we don't have critical edges in OSSA anymore this is not needed. But I added an assert and it fired in a simple control flow with frontier being at the return instruction. I'll look at this further.

eeckstein replied:
Ah yes, I didn't realize that we don't have critical edges in OSSA anymore.

eeckstein commented on lib/SILOptimizer/Transforms/RedundantLoadElimination.cpp:
```diff
@@ -1523,6 +1662,64 @@ void RLEContext::processBasicBlocksForRLE(bool Optimistic) {
+            });
+      }
+    } else {
+      // If the LastUser was a SILPhiArgument, check if any of the incoming
```
If a frontier instruction is the first in a block, it does not necessarily mean that the last user is a terminal instruction in the predecessor block. E.g.
```
  %1 = def
loop_header:
    ... = use(%1)
   cond_br loop_header, loop_exit
loop_exit:
   // This is the frontier
```

meg-gupta replied:
I missed this comment. 

Thanks for the example. But can such a use ever be consuming ?
In this non-consuming case, HasLifetimeEnd will be false (because we did not find it as an incoming value to the phi arg) and we will end up with a destroy_value correctly in the loop_exit.

eeckstein replied:
You are right, the code will work. But at least the comment is confusing, because the block might not have any phi arguments at all.

meg-gupta replied:
Ok, I fixed the comment.

eeckstein reviewed (commented):
Some initial comments

meg-gupta commented on lib/SILOptimizer/Transforms/RedundantLoadElimination.cpp:
```diff
@@ -164,6 +165,7 @@ static bool isRLEInertInstruction(SILInstruction *Inst) {
   case SILInstructionKind::EndAccessInst:
   case SILInstructionKind::SetDeallocatingInst:
   case SILInstructionKind::DeallocRefInst:
+  case SILInstructionKind::CopyValueInst:
```
Thanks. Will do.

meg-gupta commented on lib/SILOptimizer/Transforms/RedundantLoadElimination.cpp:
```diff
@@ -700,6 +727,59 @@ SILValue BlockState::reduceValuesAtEndOfBlock(RLEContext &Ctx, LSLocation &L) {
+      TheForwardingValue = createCopyAndSetState(TheForwardingValue);
+    }
+    // Mark as forwarded
+    ForwardedValues.insert(TheForwardingValue);
```
I'll add a comment explaining this better. There are two cases here: 1. a copy_value was already forwarded, then create a new copy_value 2. copy_value was not forwarded. In both the cases, we have to mark the copy_value as forwarded (either the old one or the newly created one).

meg-gupta commented on include/swift/SILOptimizer/Utils/LoadStoreOptUtils.h:
```diff
@@ -258,7 +258,22 @@ class LSValue : public LSBase {
+      // to a load
+      Res =
+          SILBuilderWithScope(Inst).emitCopyValueOperation(Inst->getLoc(), Res);
+      SILBuilderWithScope(Inst).emitEndBorrowOperation(Inst->getLoc(), Val);
```
Only if we created a begin_borrow, we need to insert end_borrow and copy_value. Am using Val != Base to combine multiple conditions.

zoecarver replied:
Ah, I see what you did :) Makes sense now, thanks.

meg-gupta commented on lib/SILOptimizer/Transforms/RedundantLoadElimination.cpp:
```diff
@@ -1637,6 +1853,32 @@ bool RLEContext::run() {
+  for (auto *Copy : NewCopies) {
+    // If a copy's use is empty, it implies we never used it to forward the
+    // loaded/stored value. Delete this copy
+    if (Copy->use_empty()) {
```
Explained in the comment above, if a load value or a stored value is never forwarded, copy_value's use will be empty. Explained why we need to create copy_value proactively here - https://github.com/apple/swift/pull/34996#discussion_r538736439

zoecarver replied:
Thanks for explaining. Finally had a chance to go read the whole thread/comments. SGTM.

meg-gupta commented on lib/SILOptimizer/Transforms/RedundantLoadElimination.cpp:
```diff
@@ -1523,6 +1666,66 @@ void RLEContext::processBasicBlocksForRLE(bool Optimistic) {
+    if (LastUser) {
+      // If the LastUser is a begin_borrow, it is not lifetime ending, change
+      // the InsertPt to next instruction after end_borrow
+      if (isa<BeginBorrowInst>(LastUser)) {
```
begin_borrow could be the last user of a value, then its next instruction will be the frontier

meg-gupta commented on lib/SILOptimizer/Transforms/RedundantLoadElimination.cpp:
```diff
@@ -700,6 +727,63 @@ SILValue BlockState::reduceValuesAtEndOfBlock(RLEContext &Ctx, LSLocation &L) {
+  // memory is not available, but dominating values of all its fields are
+  // available. Then, LSValue::reduce will create a struct from values of all
+  // its available fields.
+  assert(isa<StructInst>(TheForwardingValue));
```
Not sure when it could be an enum ?

zoecarver replied:
Maybe
```
%y = unchecked_take_enum_data_addr %x
%z = load [take] %y
```
? I think that counts as a projection. (I really like the way this is now implemented, FWIW. Handling tuples, no assertion, etc.)

meg-gupta replied:
structs and tuples are "composite" types, if all loads of their "elements" are available on all patchs, we can construct a struct/tuple from the individually loaded elements. enums don't work this way, they can only be a certain "case". https://github.com/apple/swift/blob/fd1fd2576e5eba6885278a45b16d2fbe67211abd/lib/SIL/Utils/Projection.cpp#L822 handles creating only structs and tuples from their first level projections for the same reason

zoecarver replied:
OK. I see. Thanks for explaining :)

meg-gupta commented on lib/SILOptimizer/Transforms/RedundantLoadElimination.cpp:
```diff
@@ -1523,6 +1599,194 @@ void RLEContext::processBasicBlocksForRLE(bool Optimistic) {
+  Vla.computeFrontier(ValFrontier, ValueLifetimeAnalysis::DontModifyCFG);
+  // Go over all the frontiers and insert a destroy of the copy if a
+  // lifetime ending instruction was not found at the frontier
+  for (auto *End : ValFrontier) {
```
@gottesmm I could not use the `findJointPostDominatingSet` utility here. I have some copies in RLE that will not have any consuming use. And I need to figure out the correct BasicBlock::iterator to insert the destroy_value.

meg-gupta commented:
cc @gottesmm @atrick

gottesmm commented on lib/SILOptimizer/Transforms/RedundantLoadElimination.cpp:
```diff
@@ -1596,6 +1860,11 @@ bool RLEContext::run() {
 
+  if (!Fn->hasOwnership()) {
+    // Make sure these are filled only in ossa
+    assert(NewCopies.empty() && NewPhiArgs.empty() && ForwardedValues.empty());
```
```
assert(Fn->hasOwnership() || (NewCopies.empty() && NewPhiArgs.empty() && ForwardedValues.empty()));
```

meg-gupta commented on lib/SILOptimizer/Transforms/RedundantLoadElimination.cpp:
```diff
@@ -882,6 +971,19 @@ void BlockState::processWrite(RLEContext &Ctx, SILInstruction *I, SILValue Mem,
 
   // Are we computing available value or performing RLE?
```
Done

meg-gupta commented:
@swift-ci test

meg-gupta commented:
@swift-ci benchmark

eeckstein reviewed (approved):
LGTM!

I'm just wondering if needsAdditionalCopyBeforeRedundantLoad and fixupLifetimeOfOwnedValue could be the same thing. If you run the VLA on the original load/store you get the info whether to copy and the info where to terminate the lifetime. But maybe I'm missing something here.

zoecarver commented on lib/SILOptimizer/Utils/InstOptUtils.cpp:
```diff
@@ -1388,6 +1392,91 @@ static bool keepArgsOfPartialApplyAlive(PartialApplyInst *pai,
+  ValueLifetimeAnalysis::Frontier valFrontier;
+  vla.computeFrontier(valFrontier, ValueLifetimeAnalysis::DontModifyCFG);
+
+#if NDEBUG
```
I think you want `#if !NDEBUG` or `#ifndef`. Also please put a commend on the `#endif`.

zoecarver reviewed (approved):
This LGTM. Just two more comments:
1. There are still many `TitleCase` variables. I'll stop asking about it now :P
2. You might want some of these functions in DSE, so just keep in mind that you might need to make them public APIs at some point.

meg-gupta commented:
@eeckstein That's a great point. It is true for all forwarded copies. But not true for proactive copies.
 I am re-writing this to separate the type of copies to be more clear.
 Also there is a new JointPostDominatorSet utility that was merged in recently. It makes it much simpler to write this instead of ValueLifetimeAnalysis.

meg-gupta commented:
@zoecarver I'll clang-tidy my diff to fix style

meg-gupta committed:
Fix an edge case in JointPostDominanceSetComputer::findJointPostDominatingSet

While computing blocksThatLeakIfNeverVisited by processing the worklist,
we should also add all the successors of a predecessor block that was in
the dominatedBlockSet initially. If not we will end up with incorrect
result for a control flow like this :

dominating block : bb1
dominated block : bb2

                         +----+
                         | bb1|
                         +----+
             +------------>|
             |             v
             |           +-+--+
             |           | bb2| -----+
             |           +-+--+      |
             |             |         |
             |             v         v
             |           +-+--+    +-+--+
             |           |bb3 |    | bb4|
             |           +--+-+    +----+
             |              |
             |              |
             +--------------+

Without the fix blocksThatLeakIfNeverVisited will not have bb4.

The sil test for this fix is rle_redundantload_does_not_postdominate2
in redundant_load_elim_ossa_complex.sil

meg-gupta committed:
Enable RLE on OSSA

meg-gupta commented:
@swift-ci test

meg-gupta commented:
@swift-ci test

meg-gupta commented:
@swift-ci etst source compatibility

meg-gupta commented:
@swift-ci test source compatibility

meg-gupta commented:
@swift-ci benchmark

meg-gupta commented:
@swift-ci test and merge

meg-gupta commented:
@swift-ci test

meg-gupta commented:
@swift-ci test Windows platform

meg-gupta commented:
I am working on further simplifying this patch. I'll update the PR once am done.

meg-gupta commented:
@swift-ci test

meg-gupta commented:
@swift-ci test

eeckstein commented on lib/SILOptimizer/Utils/InstOptUtils.cpp:
```diff
@@ -1963,3 +1963,45 @@ SILBasicBlock::iterator swift::replaceSingleUse(Operand *use, SILValue newValue,
+
+  auto insertPt = getInsertAfterPoint(value).getValue();
+
+  // If \p value is a newly created @owned value, we don't have to create a copy
```
This check doesn't make sense: as this utility function is a general purpose utility, it should assume that the SIL is valid when calling it. An owned value without any uses is not valid SIL.
If you actually hit this case, I would be concerned.

That means: you should always copy the value.
Maybe a better name for the function would then be `makeCopiedValueAvailable` (the "...forOwnedUse" is implied by "Copied").

meg-gupta replied:
This check was needed for newly created owned values passed into the utility. 
In RLE, for example, if we have struct field loads followed by a struct load, RLE creates a struct (similarly for tuple) to be used instead of the redundant load. This case, where we have a newly created struct does not need an additional copy. (May need a control equivalent copy).

meg-gupta replied:
@eeckstein I updated the PR with 2 versions of the api `makeCopiedValueAvailable`  and `makeNewValueAvailable` (for new values without uses) now.

eeckstein reviewed (changes requested):
That looks great!
I have just one minor change request.

meg-gupta commented:
@swift-ci test

eeckstein reviewed (approved):
perfect!
Thanks!

meg-gupta commented:
@swift-ci test

meg-gupta commented:
@swift-ci test source compatibility

meg-gupta commented:
@swift-ci test OS X Platform
